RFC: Universal Asset Notation (#335)
nervosnetwork/rfcs: rfcs/0000-universal-asset-notation/0000-universal-asset-notation.md

---
Number: ""
Category: Standards Track
Status: Proposal
Author: Jan Xie, Jordan Mack, Wenchao Hu
Organization: Nervos Foundation, Cryptape
Created: 2022-04-22
---

# Universal Asset Notation

## Abstract

This document specifies a universal naming method for assets flowing on Nervos Network, including assets on CKB and assets on layer 2 networks such as Godwoken and Axon.

## Motivation

> The interoperation between Nervos Network and other crypto or traditional economy networks brings new assets to Nervos Network. The boom of assets create difficulties for developers and users - for example, different assets on different networks may accidentally use the same name, or the same asset may arrive Nervos Network through different cross-chain bridges/routes and eventually appears as different mapping assets. A proper asset notation help incorporate assets origin from multiple networks into one network without causing any confusion, thus lowers the operational risk and facilitates deveoper and user experience.

Reviewer: ```suggestion
The interoperation between Nervos Network and other crypto or traditional economy networks brings new assets to Nervos Network. The boom of assets create difficulties for developers and users - for example, different assets on different networks may accidentally use the same name, or the same asset may arrive Nervos Network through different cross-chain bridges/routes and eventually appears as different mapping assets. A proper asset notation helps incorporate assets origin from multiple networks into one network without causing any confusion, thus lowers the operational risk and facilitates developer and user experience.
```

## Definition

> The Universal Asset Notation (UAN) consists of asset notation and route notation, which in turn are build upon asset symbol, chain symbol and bridge symbol.

Reviewer 2: typo: *build* should be *built* here.

<upper-case> ::= "A" | "B" | "C" | "D" | "E" | "F" | "G" | "H" | "I" | "J" | "K" | "L" | "M" | "N" | "O" | "P" | "Q" | "R" | "S" | "T" | "U" | "V" | "W" | "X" | "Y" | "Z"
<lower-case> ::= "a" | "b" | "c" | "d" | "e" | "f" | "g" | "h" | "i" | "j" | "k" | "l" | "m" | "n" | "o" | "p" | "q" | "r" | "s" | "t" | "u" | "v" | "w" | "x" | "y" | "z"
<digit> ::= "0" | "1" | "2" | "3" | "4" | "5" | "6" | "7" | "8" | "9"
```

- Characters `|` and `.` are chosen in order to avoid confliction with `-` widely used in LP token in DeFi applications.
- `<asset>` is the first component of UAN. It is a pair of `<asset-symbol>` and `<chain-symbol>` connected by a dot `.`, e.g. `BTC.ckb` for bitcoin resides on the CKB chain, `CKB.gw` for CKByte resides on the Godwoken chain.
- `<route>` is the route through which the asset come to Nervos Network, it's the second component of UAN. `<route>` is only required when the asset's chain of residence is different from its chain of issuance. `<route>` may consist of one or many `<path>`. If it's a multi-paths route, all paths are connected by `|`.
- `<path>` is a pair of bridge symbol and chain symbol, connected by a single dot `.`.

> - `<asset-symbol>` is a upper-case abbreviation of the asset, e.g. `BTC` for bitcoin, `CKB` for CKByte.

Reviewer 3: The asset symbol should not be case restricted, since some ERC20 do have a lowercase symbol, like `stETH` (staked ETH), and some wrapped tokens, like `wETH`, `wMANA`

- `<chain-symbol>` is a lower-case abbreviation of the source chain where the asset resided before came through the path, e.g. `eth` for the Ethereum chain, `ckb` for the CKB chain. Note the source chain may be different from the issuance chain of the asset.

> - `<bridge-symbol>` is a lower-case abbreviation of the bridge through which the asset came, e.g. `fb` for assets moved by Forcebridge.

Reviewer 3: maybe "Forcebridge" should be written as "Force Bridge"

> Examples:

Reviewer: The method of translating `|` to `via` and `.` to `from` helps readers to understand the notation and read the symbol in human language. I think it worths a section here.

Author: Could you suggest the change with a PR?

```
// CKByte on the CKB chain

> CKB.ckb

Reviewer 4: So `CKB` in CKB DApps should be renamed to `CKB.ckb`? or we can make a convention that `<asset-symbol>` without any annotation indicates that it is born on the current chain backing the DApp.

// CKByte on the Godwoken chain
CKB.gw

// WBTC crossed from the Ethereum chain to the CKB chain via Forcebridge
WBTC.ckb|fb.eth

// BNB crossed from the BSC chain to the CKB chain via Forcebridge
BNB.ckb|fb.bsc

// ETH crossed from the Ethereum chain to the CKB chain via Anotherbridge
ETH.ckb|ab.eth

> // ETH crossed from the Ethereum chain to the Godwoken chain via Forcebridge and Godwoken bridge

Reviewer 5: ```suggestion
// ETH crossed from the Ethereum chain to the Godwoken chain via Force Bridge and Godwoken Bridge
```

Reviewer 5: Follow the name style of Force Bridge

## References:
- https://github.com/nervosnetwork/force-bridge/pull/373#discussion_r924029488
- https://github.com/nervosnetwork/rfcs/pull/335/files#r913060381

> ETH.gw|gw.ckb|fb.eth

Reviewer 5: ```suggestion
ETH.gw|gb.ckb|fb.eth
```

Reviewer 5: gw: Godwoken
gb: Godwoken Bridge

- Omit the intermedia paths to emphasis the chain of origin and reduce interference, e.g. use `ETH|fb.eth` for `ETH` crossed from the Ethereum chain and eventually resides on the Godwoken chain, on the UI of a Godwoken dapp.

Another possible format of UAN display could be:

```
<asset-symbol> " (via " <bridge-name> " from " <chain-symbol-in-upper-case> ")"
```

Where `<bridge-name>` is the full name of the bridge in the rightmost path of route, and `<chain-symbol-in-upper-case>` is the chain symbol of the chain in the rightmost path of route. The chain of residence is omitted in display name because it should be clear in the context of the dapp with which a user interacts. The intermediate paths are omitted to make display name more readable.

Examples:

```
// WBTC crossed from the Ethereum chain to the CKB chain via Forcebridge
// UAN
WBTC.ckb|fb.eth
// Display Name
WBTC (via Forcebridge from ETH)

// BNB crossed from the BSC chain to the CKB chain via Forcebridge
// UAN
BNB.ckb|fb.bsc
// Display Name
BNB (via Forcebridge from BSC)

// ETH crossed from the Ethereum chain to the CKB chain via Anotherbridge
// UAN

> ETH.ckb|eth.ab

Reviewer 5: ```suggestion
ETH.ckb|ab.eth
```

Reviewer 5: `<path> ::= <bridge-symbol> "." <chain-symbol>`

// Display Name
ETH (via Anotherbridge from ETH)

> // ETH crossed from the Ethereum chain to the Godwoken chain via Forcebridge and Godwoken bridge

Reviewer 5: ```suggestion
// ETH crossed from the Ethereum chain to the Godwoken chain via Force Bridge and Godwoken Bridge
```

Reviewer 5: Follow the name style of Force Bridge

## References:
- https://github.com/nervosnetwork/force-bridge/pull/373#discussion_r924029488
- https://github.com/nervosnetwork/rfcs/pull/335/files#r913060381

// UAN

> ETH.gw|ckb.gw|eth.fb

Reviewer 5: ```suggestion
ETH.gw|gw.ckb|fb.eth
```

Reviewer 5: `<path> ::= <bridge-symbol> "." <chain-symbol>`

> ETH.gw|gw.ckb|fb.eth

Reviewer 5: ```suggestion
ETH.gw|gb.ckb|fb.eth
```

// Display Name
ETH (via Forcebridge from ETH)
```

## References

> - [Forcebridge Asset List](https://github.com/nervosnetwork/force-bridge/blob/main/configs/all-bridged-tokens.json)

Reviewer: It's weird to put a reference which does not follow the notation here.

Author: I'll make it clear this RFC is an extension of forcebridge's convention.
